File: src/view/GUI.java
package view;

import castle.Game;
import util.Echoer;
import util.MessageHandler;

import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.File;

import static javafx.scene.input.KeyCode.F;
import static javafx.scene.input.KeyCode.X;

/**
 * 视图
 * Created by asus1 on 2016/1/31.
 */
public class GUI extends Game
		implements Echoer {

	private JTextField textField;
	private JTextArea textArea;
	private JFrame frame;
	private JScrollPane scrollPane;

	private static int FRAME_X = 500;
	private static int FRAME_Y = 500;
	private static int INPUT_Y = 500;

	public GUI() {
		frame = new JFrame("城堡游戏   by 千里冰封");
		textField = new JTextField("在这里输入指令");
		textField.registerKeyboardAction(e -> {
					HandleMessage(textField.getText());
					textField.setText("");
				},
				KeyStroke.getKeyStroke(KeyEvent.VK_ENTER, 0, true),
				JComponent.WHEN_FOCUSED
		);
		textArea = new JTextArea();
		textArea.addMouseListener(new MouseListener() {
			@Override
			public void mouseClicked(MouseEvent e) {

			}
			@Override
			public void mousePressed(MouseEvent e) {}
			@Override
			public void mouseReleased(MouseEvent e) {}
			@Override
			public void mouseEntered(MouseEvent e) {}
			@Override
			public void mouseExited(MouseEvent e) {}
		});
		textArea.setBackground(new Color(12, 29, 39));
		textArea.setForeground(new Color(151, 212, 30));
		frame.setIconImage(Toolkit.getDefaultToolkit().createImage(
				"." + File.separator + "drawable" + File.separator + "ic_launcher.png"
		));
		frame.setSize(FRAME_X, FRAME_Y);
		// 绝对布局
		// frame.setLayout(null);
		frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
		frame.add(textField, BorderLayout.SOUTH);
		scrollPane = new JScrollPane(textArea);
		frame.add(scrollPane, BorderLayout.CENTER);
		frame.setResizable(false);
		frame.setLocation(FRAME_X / 8,FRAME_Y / 8);
		frame.setVisible(true);
	}

	public static void main(String[] args) {
		GUI game = new GUI();
		game.onStart();
	}

	@Override
	public void echo(String words){
//		System.out.print(words);
		textArea.append(words);
		int i = textArea.getText().length();
		int MAX_LENGTH = 10000;
		if(i > MAX_LENGTH){
			textArea.setText(textArea.getText().substring(
					i - MAX_LENGTH, i
			));
		}
	}

	@Override
	public void echoln(String words){
		echo(words + "\n");
	}

	@Override
	public void closeScreen() {
		frame.dispose();
	}

}

File: src/view/GUI.kt
package view

import castle.Game
import util.Echoer

import javax.swing.*
import java.awt.*
import java.awt.event.*
import java.io.File

/**
 * 视图
 * Created by asus1 on 2016/1/31.
 */
class GUI : Game(), Echoer {

    private val textField: JTextField
    private val textArea: JTextArea
    private val frame: JFrame
    private val scrollPane: JScrollPane
    private val scrollBar: JScrollBar

    init {
        frame = JFrame("城堡游戏   by 千里冰封")
        textField = JTextField("在这里输入指令")
        textField.registerKeyboardAction({ e ->
            HandleMessage(textField.text)
            textField.text = ""
        },
                KeyStroke.getKeyStroke(KeyEvent.VK_ENTER, 0, true),
                JComponent.WHEN_FOCUSED)
        textArea = JTextArea()
        textArea.addMouseListener(object : MouseListener {
            override fun mouseClicked(e: MouseEvent) {
            }
            override fun mousePressed(e: MouseEvent) {
            }
            override fun mouseReleased(e: MouseEvent) {
            }
            override fun mouseEntered(e: MouseEvent) {
            }
            override fun mouseExited(e: MouseEvent) {
            }
        })
        textArea.background = Color(12, 29, 39)
        textArea.foreground = Color(203, 212, 188)
        textArea.isEditable = false
        frame.iconImage = Toolkit.getDefaultToolkit().createImage(
                "." + File.separator + "src" + File.separator + "drawable" + File.separator + "ic_launcher.png")
        frame.setSize(FRAME_X, FRAME_Y)
        // 绝对布局
        // frame.setLayout(null);
        frame.defaultCloseOperation = WindowConstants.EXIT_ON_CLOSE
        frame.add(textField, BorderLayout.SOUTH)
        scrollPane = JScrollPane(textArea)
        scrollBar = scrollPane.verticalScrollBar
        frame.add(scrollPane, BorderLayout.CENTER)
        frame.isResizable = false
        frame.setLocation(FRAME_X / 8, FRAME_Y / 8)
        frame.isVisible = true
    }

    override fun echo(words: String) {
        //		System.out.print(words);
        textArea.append(words)
        val i = textArea.text.length
        val MAX_LENGTH = 10000
        if (i > MAX_LENGTH) {
            textArea.text = textArea.text.substring(
                    i - MAX_LENGTH, i)
        }
        //		scrollBar.setValue(scrollBar.getMaximum() - 20);
        // 滚动到最底下
        val height = 10
        val p = Point()
        p.setLocation(0, this.textArea.lineCount * height)
        this.scrollPane.viewport.viewPosition = p
    }

    override fun echoln(words: String) {
        echo(words + "\n")
    }

    override fun closeScreen() {
        frame.dispose()
    }

    companion object {

        private val FRAME_X = 500
        private val FRAME_Y = 500

		@JvmStatic fun main(args: Array<String>) {
            val game = GUI()
            game.onStart()
        }
    }

}
